analyzer/src/main/kotlin/managers/Cargo.kt

```diff
@@ -0,0 +1,290 @@
+            val hash = Hash.create(hashes[checksum] ?: "")
+            return RemoteArtifact(url, hash)
+        } else {
+            return null
+        }
+    }
+
+    private fun extractPackage(node: JsonNode, hashes: Map<String, String>): Package {
+        val vcsInfo = extractVcsInfo(node)
+        return Package(
+            id = extractPackageId(node),
+            declaredLicenses = extractDeclaredLicenses(node),
+            description = node["description"].textValueOrEmpty(),
+            binaryArtifact = RemoteArtifact.EMPTY,
+            sourceArtifact = extractSourceArtifact(node, hashes) ?: RemoteArtifact.EMPTY,
+            homepageUrl = "",
+            vcs = vcsInfo,
+            vcsProcessed = vcsInfo.normalize()
+        )
+    }
+
+    private fun checksumKeyOf(metadata: JsonNode): String {
+        val id = metadata["id"]!!.textValue()
+        return "\"checksum $id\""
+    }
+
+    // Cargo.lock is located next to Cargo.toml or in one of the parent directories. The latter
+    // is case when the project is part of a workspace. Cargo.lock is then located next to the
+    // Cargo.toml file defining the workspace.
+    private fun findCargoLock(definitionFile: File): File {
+        var workingDir = definitionFile.parentFile
+        while (workingDir != analyzerRoot.parentFile) {
+            val lockfile = File(workingDir, "Cargo.lock")
+            if (lockfile.exists()) {
+                return lockfile
+            }
+            workingDir = workingDir.parentFile
+        }
+        // we reached the root directory we are analyzing
+        throw IllegalArgumentException("missing Cargo.lock file")
+    }
+
+    private fun readHashes(lockfile: File): Map<String, String> {
+        val contents = Toml().read(lockfile)
+        val metadata = contents.getTable("metadata") ?: return emptyMap()
+        val metadataMap = metadata.toMap()
+        return metadataMap.asSequence().mapNotNull {
+            val key = it.key
+            val value = it.value as? String
+            if (key != null && value != null) {
+                Pair(key, value)
+            } else {
+                null
+            }
+        }.toMap()
+    }
+
+    private fun resolveDependenciesTree(
+        metadata: JsonNode,
+        packages: Map<String, Package>,
+        filter: (pkgId: String, depId: String) -> Boolean = { _, _ -> true }
+    ): SortedSet<PackageReference> {
+        val resolve = metadata["resolve"]
+        val nodes = resolve["nodes"]
+        val root = resolve["root"].textValueOrEmpty()
+
+        return resolveDependenciesOf(
+            root, nodes, packages, filter
+        ).dependencies
+    }
+
+    private fun resolveDependenciesOf(
+        id: String,
+        nodes: JsonNode,
+        packages: Map<String, Package>,
+        filter: (pkgId: String, depId: String) -> Boolean
+    ): PackageReference {
+        val node = nodes.find { it["id"].textValueOrEmpty() == id }!!
```

Comment:
If you're sure there's exactly one match, use `.single {}` instead of `.find {}!!`.